// @olduvai-jp/pixi-animated-webp/lib/AnimatedWebPAsset.js
'use strict';

var pixi_js = require('pixi.js');
var AnimatedWebP = require('./AnimatedWebP.js');

const AnimatedWebPAsset = {
  extension: pixi_js.ExtensionType.Asset,
  detection: {
    test: async () => true,
    add: async (formats) => [...formats, "webp"],
    remove: async (formats) => formats.filter((format) => format !== "webp")
  },
  loader: {
    name: "animatedWebpLoader",
    extension: {
      type: pixi_js.ExtensionType.LoadParser,
      priority: 10
    },
    test: (url) => {
      const isWebP = pixi_js.path.extname(url) === ".webp";
      return isWebP;
    },
    load: async (url, asset) => {
      try {
        const response = await pixi_js.DOMAdapter.get().fetch(url);
        const buffer = await response.arrayBuffer();
        const result = await AnimatedWebP.AnimatedWebP.fromBuffer(buffer, asset?.data);
        return result;
      } catch (error) {
        console.error("Error in AnimatedWebPAsset loader:", error);
        throw error;
      }
    },
    unload: async (asset) => {
      asset.destroy();
    }
  }
};
pixi_js.extensions.add(AnimatedWebPAsset);

exports.AnimatedWebPAsset = AnimatedWebPAsset;
